### DISCUSSION TITLE
A new way to check whether an element is included in a collection(like:"a".IsIn(a collection))

### DISCUSSION THREAD

**Discussion by hrx521** [dotnet/csharplang#4255] [Language Ideas] [answered]
I hope the C# language adds a new method to the type system, which is used to check whether the specified element is contained in a collection. This is different ways of expression from the existing collection' method "contains". 

At present, although a collection has a "contains" method to check whether it contains a specified element, the semantics of the two are quite different. If we can check it in the following ways,the relationship between subject and object can be expressed more clearly. After all, we always want to use computer language in a way that is closer to the way human think.
For the details of this design, I can think of some suggestions at present:

When comparing relationships contained in, sometimes it is to compare whether the value contains or not, and sometimes to compare references. Both should be achieved.
for example, the following pseudo code:
```
//1. For value types,comparison value
int[] myValueArray = { 5, 9, 15 };
if (3.IsIn(myValueArray)) { }
//or as the follow forms 
if (5 IsIn myValueArray) { }

//2. For reference types, Reference comparison as default
var strList = new string[] { "a", "b" };
object x = "b";
if (x.IsIn(strList)){ }
if (x.IsInAsValued(strList)){ }// Specify compared by value

//or as the follow forms
if (x IsIn strList){ }
if (x IsInAsValued strList){ }// Specify compared by value
```

I can't think of more details for the moment, and I hope the experts can discuss more.

**Comment by Phantonia** [comment, marked as answer]
I don't know if I misunderstand you, but that's the matter of a single extension method:
```cs
using System.Collections.Generic;

public static class Extensions
{
    public static bool IsIn<T>(this T item, ICollection<T> collection) => collection.Contains(item);
}
```
If the extension method is in scope, the syntax would be exactly as you mentioned:
```cs
int[] myValueArray = { 5, 9, 15 };
if (3.IsIn(myValueArray)) { }
```
And for `IsInAsValued` I don't think there is a general answer to how to implement that, so I suppose it would be necessary for you to write it yourself anyway to fulfill your own needs.

**Reply by hrx521** [reply]
Thank you for participating in this discussion

**Comment by YairHalberstadt** [comment]
Do you think any language changes are needed here? It looks to me like this a base class library request, which are taken on dotnet/runtime

**Reply by hrx521** [reply]
Thank you for participating in this discussion

**Comment by huoyaoyuan** [comment]
> we always want to use computer language in a way that is closer to the way human think

I'm always careful with words like this.

The way of human think varies for such simple expressions, and can be influenced by their mother language. Although I have the same language with you, I don't think there is a significant difference for the order.

An important part of C# design philosophy should be the **inference experience of IDE**. Here the collection variable is typed first, so the inference of "contain" is more natural. "Contain" is a common operation for collections, but it's not that common for all objects in total.

**Reply by hrx521** [reply]
Thank you for participating in this discussion
